wordpress-mobile/WordPress-Android
# Xmlrpc login errors.

Fixes #12401

Needed PR to merge before: 
- [Fluxc](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2280)
- [Login Lib](https://github.com/wordpress-mobile/WordPress-Login-Flow-Android/pull/78)

This PR aims to improve error messages that can occur during login for self-hosted sites (see [this comment](https://github.com/wordpress-mobile/WordPress-Android/issues/12401#issuecomment-1021615025) for more details).

**NOTE 1:** for both FluxC and Login libs I tried to keep the current behaviors only adding a couple of enums mapping the various errors conditions for xml-rpc, that are used/parsed only in the login flow for self-hosted. This should reduce the possibility of regressions but appreciate testing of the various log-in/sign-up scenarios 🙇 .

**NOTE 2:** hi @hichamboushaba 👋 , as for another recent PR set, also this one implies a few changes in the FluxC and Login libs (see linked PRs at the beginning of this description). Not sure if they could have an impact on Woo app, but maybe we can sync the testing/merge of this PRs set with the testing/merge of [this other PR](https://github.com/wordpress-mobile/WordPress-Login-Flow-Android/pull/76) (and related fluxc) so to have only 1 PR for Woo. Let me know 🙇 

## To test
As per  [this comment](https://github.com/wordpress-mobile/WordPress-Android/issues/12401#issuecomment-1021615025) install the [xml-rpc-tweak.zip](https://github.com/wordpress-mobile/WordPress-Android/files/7937160/xml-rpc-tweak.zip) plugin in a self-hosted site; once done that you will have a settings page for the plugin in the wp-admin that allows to set the various test MODEs. Here below the expected results while logging in:

| MODE | Error view |
|---|---|
| Mode 0 / Off - XML-RPC is working, but a bad username / password was provided. | ![image](https://user-images.githubusercontent.com/47797566/153299911-220e6549-4a2e-4893-a40b-98dbeedd7827.png) |
| Mode 1 - Authenticated XML-PRC requests are disabled. | ![image](https://user-images.githubusercontent.com/47797566/153300090-c1fa8937-be2d-487b-b3bf-a641cb2a4aba.png)|
| Mode 2 - Simulates a server that responded with an invalid payload the client couldn't decode. | ![image](https://user-images.githubusercontent.com/47797566/153300219-8602391f-4fb2-4719-887b-f7c85d747b8f.png) |
| Mode 3 - Simulates a server that responded with a bad HTTP code and no payload. | ![image](https://user-images.githubusercontent.com/47797566/153300561-9f4272c8-e3fa-477c-8f50-d51e59bb32d2.png) | 
| Mode 4 - Block all XML-RPC calls. | ![image](https://user-images.githubusercontent.com/47797566/153300964-c75fa568-6a42-4563-924b-79f13ba5756c.png) | 

### Smoke test login/signup
Using wpcom/jp/self-hosted/atomic sites smoke test the various login/signup flows. Test also magic links, 2FA, reset password, or anything else you could think of 🙇 .

## Regression Notes
1. Potential unintended areas of impact
Based on how the modifications were done, effects could be in case present only on self-hosted sites.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing.

3. What automated tests I added (or what prevented me from doing so)
Autometed tests for login flows.

PR submission checklist:
